src/test/java/MacroExecutionModuleImplementationTest.java

```diff
@@ -0,0 +1,236 @@
+    }
+
+    // Test sending a message in thread where a macro has already been used and you are the creator.
+    @Test
+    public void useMacroAlreadyUsedInThreadAsTheCreator() throws IOException, GeneralSecurityException {
+        addToThreadMacroMap(THREAD1, MACRO_NAME);
+
+        setUpMocks();
+        String expectedText = createDefaultChatResponse(USER_EMAIL1);
+
+        ChatResponse chatResponse = execution.getReplyText(MESSAGE_WITHOUT_MACRO_NAME, THREAD1, ROOM1, USER_EMAIL1, HELP_MESSAGE);
+        String actual = chatResponse.getReplyText();
+
+        Assert.assertEquals(expectedText, actual);
+    }
+
+    // Test sending a message in a thread where a macro has already been used and the macro has been shared (you are not the creator).
+    @Test
+    public void useSharedMacroAlreadyUsedInThread() throws IOException, GeneralSecurityException {
+        addToThreadMacroMap(THREAD1, MACRO_NAME);
+        shareDefaultMacroToRoom(ROOM1);
+
+        setUpMocks();
+        String expectedText = createDefaultChatResponse(USER_EMAIL2);
+
+        ChatResponse chatResponse = execution.getReplyText(MESSAGE_WITH_MACRO_NAME, THREAD1, ROOM1, USER_EMAIL2, HELP_MESSAGE);
+        String actual = chatResponse.getReplyText();
+
+        Assert.assertEquals(expectedText, actual);
+    }
+
+    // Test using the share message.
+    @Test
+    public void testShareMessage() throws IOException, GeneralSecurityException {
+        setUpMocks();
+        
+        ChatResponse chatResponse = execution.getReplyText(SHARE_MESSAGE, THREAD1, ROOM1, USER_EMAIL1, HELP_MESSAGE);
+        String actual = chatResponse.getReplyText();
+
+        Assert.assertEquals(SHARE_MESSAGE_REPLY, actual);
+    }
+
+    private void addToThreadMacroMap(String thread, String macroName) {
+        execution.getThreadMacroMap().put(thread, macroName);
+    }
+
+    private void shareDefaultMacroToRoom(String room) {
+        Map<String, String> macroToCreator = new HashMap<String, String> ();
+        macroToCreator.put(DEFAULT_MACRO.getMacroName(), DEFAULT_MACRO.getCreatorId());
+        execution.getRoomToMacro().put(room, macroToCreator);
+    }
+
+    private String getDate(String pattern) {
+        // Create timestamp based on pattern provided
+        LocalDateTime myDateObj = LocalDateTime.now();
+        DateTimeFormatter myFormatObj = DateTimeFormatter.ofPattern(pattern);
+        return myDateObj.format(myFormatObj);
+    }
+
+    private String createDefaultChatResponse(String userEmail) {
+        List<String> values = new ArrayList<String>();
+        values.add(getDate("dd-MM-yyyy HH:mm:ss"));
+        values.add(userEmail);
+        values.add(MESSAGE_CONTENT_ONLY);
+        return (ChatResponse.createChatResponseWithList(values, DEFAULT_MACRO.getAction().getActionType(), DEFAULT_MACRO.getAction().getDocumentUrl())).getReplyText();
+    }
+
+    private void setUpMocks() throws IOException, GeneralSecurityException {
```

Comment:
super nit: the name of this method should be more descriptive - what exactly are you setting up?